substrate/primitives/core/src/const_hex2array.rs

```diff
@@ -0,0 +1,162 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides a const function for converting a hex string to a `u8` array at compile time, when used
+//! in the proper context.
+
+/// Provides a const array from given string literal.
+///
+/// Valid characters are `[0-9a-fA-F]`, and the hex string should not start
+/// with the `0x` prefix.
+#[macro_export]
+macro_rules! hex2array {
+	($input:expr) => {{
+		const BYTES: [u8; $input.len() >> 1] = $crate::const_hex2array::private_hex2array($input);
+		BYTES
+	}};
+}
+
+/// Generates array from (static) string literal.
+///
+/// Valid characters are `[0-9a-fA-F]`, and the hex string should not start
+/// with the `0x` prefix.
+///
+/// # Panics
+///
+/// The function will panic at compile time when used in a const context if:
+/// - The given hex string has an invalid length.
+/// - It contains invalid characters.
+///
+/// The function will panic at runtime when used in a non-const context if the above conditions are
+/// met.
+#[doc(hidden)]
+pub const fn private_hex2array<const N: usize>(hex: &str) -> [u8; N] {
+	const fn c2b(c: u8) -> u8 {
+		match c as char {
+			'0'..='9' => c - 48,
+			'a'..='f' => c - 87,
+			'A'..='F' => c - 55,
```

Comment:
Slightly more readable version without the magic constants?

```suggestion
			'0'..='9' => c - b'0',
			'a'..='f' => c - (b'a' - 10),
			'A'..='F' => c - (b'A' - 10),
```